File: backend/10-wallet/internal/app/openapi.go
package app

import "github.com/dokandar/dokandar-wallet/internal/config"

// openapiSpec is the hand-built OpenAPI 3.0.3 document. CI invariant: every
// served REST route appears here (Go is a hand-spec stack). info.version is the
// CODE_VERSION value injected by the caller via config.CodeVersion().
func openapiSpec(serviceName string) map[string]any {
	jsonObj := func(props map[string]any, required ...string) map[string]any {
		m := map[string]any{"type": "object", "properties": props}
		if len(required) > 0 {
			m["required"] = required
		}
		return m
	}
	strProp := map[string]any{"type": "string"}

	walletSchema := jsonObj(map[string]any{
		"user_id":         map[string]any{"type": "string", "format": "uuid", "description": "opaque auth user id (the wallet owner)"},
		"currency":        map[string]any{"type": "string", "description": "ISO-4217; always BDT", "example": "BDT"},
		"status":          map[string]any{"type": "string", "enum": []string{"active", "frozen", "closed"}, "description": "wallet lifecycle state"},
		"balance_minor":   map[string]any{"type": "integer", "format": "int64", "description": "ledger balance in paisa (minor units); capped at 5,000,000 (=50,000 BDT)"},
		"available_minor": map[string]any{"type": "integer", "format": "int64", "description": "balance minus holds, in paisa"},
		"version":         map[string]any{"type": "integer", "format": "int64", "description": "optimistic-lock row version"},
	})
	entrySchema := jsonObj(map[string]any{
		"id":              map[string]any{"type": "string", "format": "uuid"},
		"wallet_user_id":  map[string]any{"type": "string", "format": "uuid"},
		"debit_minor":     map[string]any{"type": "integer", "format": "int64", "description": "paisa debited (XOR with credit_minor)"},
		"credit_minor":    map[string]any{"type": "integer", "format": "int64", "description": "paisa credited (XOR with debit_minor)"},
		"kind":            map[string]any{"type": "string", "description": "ledger entry kind", "example": "cashback"},
		"order_id":        map[string]any{"type": "string", "format": "uuid", "nullable": true, "description": "originating order id, if any"},
		"idempotency_key": map[string]any{"type": "string", "description": "unique key making at-least-once effectively-once"},
		"posted_at":       map[string]any{"type": "string", "format": "date-time", "description": "RFC-3339 UTC"},
	})
	ruleSchema := jsonObj(map[string]any{
		"id":                 map[string]any{"type": "string", "format": "uuid"},
		"trigger":            map[string]any{"type": "string", "description": "event that fires the rule", "example": "order_delivered"},
		"funded_by":          map[string]any{"type": "string", "description": "who funds the reward (platform/seller)", "example": "platform"},
		"reward_kind":        map[string]any{"type": "string", "enum": []string{"percent", "flat"}, "description": "percent of subtotal or a flat paisa amount"},
		"reward_value":       map[string]any{"type": "integer", "format": "int64", "description": "basis points if percent, else paisa"},
		"reward_cap_minor":   map[string]any{"type": "integer", "format": "int64", "nullable": true, "description": "max reward in paisa"},
		"min_subtotal_minor": map[string]any{"type": "integer", "format": "int64", "nullable": true, "description": "min order subtotal in paisa to qualify"},
		"max_per_user":       map[string]any{"type": "integer", "description": "lifetime cap per user"},
	})

	// Request-body schemas (reused as components, $ref'd from operations).
	topupReqSchema := jsonObj(map[string]any{
		"amount_minor": map[string]any{"type": "integer", "format": "int64", "minimum": 1, "description": "top-up amount in paisa (minor units)", "example": 50000},
	}, "amount_minor")
	debitReqSchema := jsonObj(map[string]any{
		"user_id":         map[string]any{"type": "string", "format": "uuid", "description": "wallet owner"},
		"amount_minor":    map[string]any{"type": "integer", "format": "int64", "minimum": 1, "description": "paisa to debit"},
		"idempotency_key": map[string]any{"type": "string", "description": "caller-supplied, makes the debit effectively-once"},
		"order_id":        map[string]any{"type": "string", "format": "uuid", "nullable": true, "description": "originating order id (optional)"},
	}, "user_id", "amount_minor", "idempotency_key")
	creditReqSchema := jsonObj(map[string]any{
		"user_id":         map[string]any{"type": "string", "format": "uuid", "description": "wallet owner"},
		"amount_minor":    map[string]any{"type": "integer", "format": "int64", "minimum": 1, "description": "paisa to credit"},
		"idempotency_key": map[string]any{"type": "string", "description": "caller-supplied, makes the credit effectively-once"},
		"kind":            map[string]any{"type": "string", "description": "credit reason (e.g. cashback, refund, reversal)", "example": "cashback"},
		"order_id":        map[string]any{"type": "string", "format": "uuid", "nullable": true, "description": "originating order id (optional)"},
	}, "user_id", "amount_minor", "idempotency_key")
	// ErrorEnvelope — the single shared platform error shape
	// {error:{code,message,request_id,details}}. $ref'd on every 4xx/5xx.
	errorSchema := jsonObj(map[string]any{
		"error": jsonObj(map[string]any{
			"code":       map[string]any{"type": "string", "description": "stable machine code (lowercase snake_case)", "example": "validation_error"},
			"message":    map[string]any{"type": "string", "description": "human-readable, scrubbed (never leaks 5xx internals)", "example": "amount_minor must be a positive integer"},
			"request_id": map[string]any{"type": "string", "description": "honour-or-mint x-request-id echo", "example": "11111111-1111-4111-8111-111111111111"},
			"details":    map[string]any{"type": "object", "description": "optional structured context", "additionalProperties": true},
		}, "code", "message", "request_id"),
	}, "error")

	ref := func(name string) map[string]any {
		return map[string]any{"$ref": "#/components/schemas/" + name}
	}
	jsonContent := func(schema map[string]any) map[string]any {
		return map[string]any{"application/json": map[string]any{"schema": schema}}
	}
	jsonContentEx := func(schema map[string]any, example map[string]any) map[string]any {
		return map[string]any{"application/json": map[string]any{"schema": schema, "example": example}}
	}
	resp := func(desc string, schema map[string]any) map[string]any {
		r := map[string]any{"description": desc}
		if schema != nil {
			r["content"] = jsonContent(schema)
		}
		return r
	}
	errResp := func(desc string) map[string]any { return resp(desc, ref("ErrorEnvelope")) }

	bearer := []map[string]any{{"bearerJwt": []any{}}}
	internal := []map[string]any{{"internalToken": []any{}}}

	// Example bodies (BDT amounts are integer paisa / minor units).
	walletEx := map[string]any{
		"user_id": "11111111-1111-4111-8111-111111111111", "currency": "BDT",
		"status": "active", "balance_minor": 150000, "available_minor": 150000, "version": 3,
	}
	topupEx := map[string]any{"amount_minor": 50000}
	debitEx := map[string]any{
		"user_id": "11111111-1111-4111-8111-111111111111", "amount_minor": 25000,
		"idempotency_key": "order-9c1f-debit-1", "order_id": "22222222-2222-4222-8222-222222222222",
	}
	creditEx := map[string]any{
		"user_id": "11111111-1111-4111-8111-111111111111", "amount_minor": 25000,
		"idempotency_key": "order-9c1f-credit-1", "kind": "cashback",
		"order_id": "22222222-2222-4222-8222-222222222222",
	}

	return map[string]any{
		"openapi": "3.0.3",
		"info": map[string]any{
			"title":   "DOKANDAR Wallet Service",
			"version": config.CodeVersion(),
			"description": "**service_name**: `" + serviceName + "` &nbsp;|&nbsp; **code_version**: `" + config.CodeVersion() + "` &nbsp;|&nbsp; " +
				"**env_version / tenant / env**: see `GET /data` identity block.\n\n" +
				"Double-entry wallet ledger: top-ups, cashback rules, and east-west debit/credit. " +
				"Money is integer **paisa** (minor units); a wallet is capped at 5,000,000 paisa (50,000 BDT).\n\n" +
				"### How to test\n" +
				"1. Click **Authorize** and paste a Bearer **access token** from the auth service " +
				"(`POST /api/v1/auth/login/request` → `/login/verify`). Customer routes (`/me`, `/me/entries`, `/me/topup`) need it.\n" +
				"2. `GET /api/v1/wallet/cashback-rules` is public (no token).\n" +
				"3. East-west routes (`/debit`, `/credit`, `/balance/{user_id}`) require the shared **`x-internal-token`** header, not a JWT.\n" +
				"4. `POST /me/topup` requires a unique **`Idempotency-Key`** header — reusing one returns the original wallet (effectively-once).",
			"contact": map[string]any{
				"name":  "DOKANDAR Platform",
				"url":   "https://dokandar.com.bd",
				"email": "api@dokandar.com.bd",
			},
			"license": map[string]any{"name": "Proprietary"},
		},
		"servers": []map[string]any{
			{"url": "https://api.dokandar.com.bd", "description": "prod"},
			{"url": "http://localhost:10010", "description": "local"},
		},
		"tags": []map[string]any{
			{"name": "ops", "description": "Operational / contract surface (/ready /health /data /metrics /docs /openapi.json)"},
			{"name": "wallet", "description": "Customer wallet: balance, ledger history, top-ups"},
			{"name": "cashback", "description": "Active cashback rules (public read)"},
			{"name": "ledger", "description": "East-west double-entry debit/credit + balance reads (x-internal-token)"},
		},
		"components": map[string]any{
			"securitySchemes": map[string]any{
				"bearerJwt": map[string]any{
					"type": "http", "scheme": "bearer", "bearerFormat": "JWT",
				},
				"internalToken": map[string]any{
					"type": "apiKey", "in": "header", "name": "x-internal-token",
				},
			},
			"schemas": map[string]any{
				"Wallet":        walletSchema,
				"Entry":         entrySchema,
				"CashbackRule":  ruleSchema,
				"TopupRequest":  topupReqSchema,
				"DebitRequest":  debitReqSchema,
				"CreditRequest": creditReqSchema,
				"ErrorEnvelope": errorSchema,
			},
		},
		"paths": map[string]any{
			"/ready": map[string]any{"get": map[string]any{
				"tags":        []string{"ops"},
				"operationId": "getReady",
				"summary":     "Readiness probe (gates on Postgres only)",
				"description": "LB/readiness gate. 200 only when Postgres (system-of-record) is reachable; degradable deps never gate.",
				"responses": map[string]any{
					"200": resp("ready", nil), "503": resp("not_ready", nil),
				},
			}},
			"/health": map[string]any{"get": map[string]any{
				"tags":        []string{"ops"},
				"operationId": "getHealth",
				"summary":     "Full diagnostics (postgres, redis, kafka, mongo_logs, elasticsearch + observability)",
				"description": "Full diagnostics over all deps plus an observability block. gRPC peer checks are diagnostic-only and never flip status.",
				"responses": map[string]any{
					"200": resp("healthy", nil), "503": resp("unhealthy", nil),
				},
			}},
			"/data": map[string]any{"get": map[string]any{
				"tags":        []string{"ops"},
				"operationId": "getData",
				"summary":     "Tenant snapshot + identity block",
				"description": "Identity block prepended to the read-only data/<tenant>/result.json snapshot (not live DB introspection).",
				"responses": map[string]any{
					"200": resp("snapshot", nil), "404": errResp("no_snapshot"), "500": errResp("snapshot_parse_failed"),
				},
			}},
			"/metrics": map[string]any{"get": map[string]any{
				"tags":        []string{"ops"},
				"operationId": "getMetrics",
				"summary":     "Prometheus exposition",
				"description": "Prometheus text exposition (RED + service counters + wallet_outbox_pending gauge). Closed-set labels only.",
				"responses":   map[string]any{"200": resp("prometheus text", nil)},
			}},
			"/docs": map[string]any{"get": map[string]any{
				"tags":        []string{"ops"},
				"operationId": "getDocs",
				"summary":     "Swagger UI",
				"responses":   map[string]any{"200": resp("html", nil)},
			}},
			"/openapi.json": map[string]any{"get": map[string]any{
				"tags":        []string{"ops"},
				"operationId": "getOpenapi",
				"summary":     "This document",
				"responses":   map[string]any{"200": resp("openapi", nil)},
			}},
			"/api/v1/wallet/me": map[string]any{"get": map[string]any{
				"tags":        []string{"wallet"},
				"operationId": "getMyWallet",
				"summary":     "Get (or auto-create) the caller's wallet",
				"description": "Returns the JWT subject's wallet, creating an empty active wallet on first read.",
				"security":    bearer,
				"responses": map[string]any{
					"200": map[string]any{"description": "wallet", "content": jsonContentEx(ref("Wallet"), walletEx)},
					"401": errResp("token_missing / token_invalid"),
				},
			}},
			"/api/v1/wallet/me/entries": map[string]any{"get": map[string]any{
				"tags":        []string{"wallet"},
				"operationId": "listMyEntries",
				"summary":     "Ledger history (newest first)",
				"description": "Paginated double-entry ledger for the caller's wallet, newest first.",
				"security":    bearer,
				"parameters": []map[string]any{{
					"name": "size", "in": "query", "required": false,
					"description": "page size (newest-first)", "example": 50,
					"schema": map[string]any{"type": "integer", "default": 50, "minimum": 1, "maximum": 200},
				}},
				"responses": map[string]any{
					"200": resp("entries", map[string]any{"type": "array", "items": ref("Entry")}),
					"401": errResp("token_missing / token_invalid"),
				},
			}},
			"/api/v1/wallet/me/topup": map[string]any{"post": map[string]any{
				"tags":        []string{"wallet"},
				"operationId": "topupMyWallet",
				"summary":     "Top up the wallet (Idempotency-Key header required)",
				"description": "Credits the caller's wallet by amount_minor (paisa). Reusing an Idempotency-Key returns the original result (effectively-once). Rejected if it would exceed the 50,000 BDT cap.",
				"security":    bearer,
				"parameters": []map[string]any{{
					"name": "Idempotency-Key", "in": "header", "required": true,
					"description": "unique per logical top-up", "example": "topup-2026-06-20-001",
					"schema": strProp,
				}},
				"requestBody": map[string]any{"required": true, "content": jsonContentEx(ref("TopupRequest"), topupEx)},
				"responses": map[string]any{
					"201": map[string]any{"description": "wallet after top-up", "content": jsonContentEx(ref("Wallet"), walletEx)},
					"400": errResp("missing_idempotency_key"),
					"409": errResp("wallet_max_exceeded"),
					"422": errResp("invalid_request"),
				},
			}},
			"/api/v1/wallet/cashback-rules": map[string]any{"get": map[string]any{
				"tags":        []string{"cashback"},
				"operationId": "listCashbackRules",
				"summary":     "Active cashback rules (public)",
				"description": "Public list of currently-active cashback rules. No authentication required.",
				"responses":   map[string]any{"200": resp("rules", map[string]any{"type": "array", "items": ref("CashbackRule")})},
			}},
			"/api/v1/wallet/debit": map[string]any{"post": map[string]any{
				"tags":        []string{"ledger"},
				"operationId": "debitWallet",
				"summary":     "Debit a wallet (east-west; x-internal-token)",
				"description": "East-west call (used by the order saga). Debits amount_minor from the user's wallet under SERIALIZABLE; idempotency_key makes the debit effectively-once. Requires x-internal-token.",
				"security":    internal,
				"requestBody": map[string]any{"required": true, "content": jsonContentEx(ref("DebitRequest"), debitEx)},
				"responses": map[string]any{
					"200": map[string]any{"description": "wallet after debit", "content": jsonContentEx(ref("Wallet"), walletEx)},
					"401": errResp("unauthorized (missing/invalid x-internal-token)"),
					"409": errResp("insufficient_balance"),
					"422": errResp("invalid_request"),
				},
			}},
			"/api/v1/wallet/credit": map[string]any{"post": map[string]any{
				"tags":        []string{"ledger"},
				"operationId": "creditWallet",
				"summary":     "Credit a wallet (east-west; x-internal-token)",
				"description": "East-west call (cashback, refund, saga compensation). Credits amount_minor to the user's wallet; idempotency_key makes the credit effectively-once. Requires x-internal-token.",
				"security":    internal,
				"requestBody": map[string]any{"required": true, "content": jsonContentEx(ref("CreditRequest"), creditEx)},
				"responses": map[string]any{
					"200": map[string]any{"description": "wallet after credit", "content": jsonContentEx(ref("Wallet"), walletEx)},
					"401": errResp("unauthorized (missing/invalid x-internal-token)"),
					"409": errResp("wallet_max_exceeded"),
					"422": errResp("invalid_request"),
				},
			}},
			"/api/v1/wallet/balance/{user_id}": map[string]any{"get": map[string]any{
				"tags":        []string{"ledger"},
				"operationId": "getWalletBalance",
				"summary":     "Read a wallet balance (east-west; x-internal-token)",
				"description": "East-west balance read by user id. Requires x-internal-token.",
				"security":    internal,
				"parameters": []map[string]any{{
					"name": "user_id", "in": "path", "required": true,
					"description": "wallet owner (auth user id)", "example": "11111111-1111-4111-8111-111111111111",
					"schema": map[string]any{"type": "string", "format": "uuid"},
				}},
				"responses": map[string]any{
					"200": map[string]any{"description": "wallet", "content": jsonContentEx(ref("Wallet"), walletEx)},
					"400": errResp("bad_request"),
					"401": errResp("unauthorized (missing/invalid x-internal-token)"),
				},
			}},
		},
	}
}

const swaggerHTML = `<!doctype html>
<html><head><title>10-wallet API</title>
<meta charset="utf-8"/>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/swagger-ui-dist@5/swagger-ui.css">
</head><body><div id="ui"></div>
<script src="https://cdn.jsdelivr.net/npm/swagger-ui-dist@5/swagger-ui-bundle.js"></script>
<script>SwaggerUIBundle({url:"/openapi.json", dom_id:"#ui", deepLinking:true, persistAuthorization:true})</script>
</body></html>`

File: backend/05-search/src/openapi.rs
//! Hand-built OpenAPI 3.0.3 + Swagger UI (mirrors 03-seller's depth: segmented tags,
//! typed component schemas, documented parameters/responses, $ref envelopes).
use crate::config::Config;
use serde_json::{json, Value};

fn ref_(name: &str) -> Value { json!({ "$ref": format!("#/components/schemas/{name}") }) }
fn json_resp(desc: &str, schema: &str) -> Value {
    json!({ "description": desc, "content": { "application/json": { "schema": ref_(schema) } } })
}

pub fn spec(cfg: &Config) -> Value {
    json!({
        "openapi": "3.0.3",
        "info": {
            "title": "DOKANDAR Search Service",
            "version": cfg.code_version,
            "description": format!(
                "**service_name**: `{}` &nbsp;|&nbsp; **code_version**: `{}` &nbsp;|&nbsp; **env_version**: `{}` &nbsp;|&nbsp; **tenant**: `{}` &nbsp;|&nbsp; **env**: `{}`\n\n\
**Discovery (CQRS read projection).** Consume-only Kafka projectors hydrate the PostgreSQL `*_view` store (+ Elasticsearch `:9201`); public reads on `/api/v1/search/*`, admin reindex behind RS256. Bilingual (Bangla/English) full-text + trigram + geo. Responses are pretty-printed JSON; errors use the standard `{{error:{{code,message,request_id,details}}}}` envelope.\n\n\
### How to test\n\
1. The discovery reads (`GET /api/v1/search/products`, `/autocomplete`, `/shops`, `/trending`, `/categories/tree`) are **public** — no token needed.\n\
2. Only `POST /api/v1/search/admin/reindex` is protected: click **Authorize** and paste a Bearer **access token** with the `admin` role minted by 01-auth (`POST /api/v1/auth/login/request` → `/login/verify`). A non-admin token returns `403`, a missing/invalid token `401`.\n\
3. `products` supports `q`, `locale` (en|bn), `category_id`, `page`, `size`; `shops` requires `lat`+`lng`. Out-of-range params return `422 invalid_request`.",
                cfg.service_name, cfg.code_version, cfg.env_version, cfg.tenant, cfg.app_env
            ),
            "contact": { "name": "DOKANDAR Platform", "url": "https://dokandar.com.bd", "email": "api@dokandar.com.bd" },
            "license": { "name": "Proprietary" }
        },
        "servers": [
            { "url": "https://api.dokandar.com.bd", "description": "prod" },
            { "url": "http://localhost:10005", "description": "local" }
        ],
        "tags": [
            { "name": "ops", "description": "Operational contract — readiness, health, data snapshot, metrics." },
            { "name": "products", "description": "Product discovery — bilingual full-text search, facets, autocomplete." },
            { "name": "shops", "description": "Shop discovery — geo radius (earthdistance) search." },
            { "name": "discovery", "description": "Trending products and the category tree." },
            { "name": "admin", "description": "Administrative operations (RS256, admin role required)." }
        ],
        "paths": {
            "/ready": { "get": { "operationId": "getReady", "tags": ["ops"], "summary": "Readiness probe", "description": "200 iff PostgreSQL (the read model) is reachable; the LB gates on this. ES/Kafka/Mongo/APM do NOT gate (search degrades to the PG tsvector fallback).", "responses": {
                "200": json_resp("ready", "ReadyResponse"), "503": json_resp("not ready", "ReadyResponse") } } },
            "/health": { "get": { "operationId": "getHealth", "tags": ["ops"], "summary": "Liveness + every dependency's health", "description": "Full diagnostics across postgres, the business-search ES (:9201), the log-sink ES (:9200), kafka, mongo, apm, plus the projection row counts.", "responses": {
                "200": json_resp("healthy", "HealthResponse"), "503": json_resp("unhealthy", "HealthResponse") } } },
            "/data": { "get": { "operationId": "getData", "tags": ["ops"], "summary": "Tenant data snapshot", "description": "The identity block prepended to the read-only `data/<tenant>/result.json` snapshot.", "responses": {
                "200": { "description": "snapshot" }, "404": json_resp("no snapshot", "ErrorEnvelope"), "500": json_resp("snapshot not an object", "ErrorEnvelope") } } },
            "/metrics": { "get": { "operationId": "getMetrics", "tags": ["ops"], "summary": "Prometheus exposition", "description": "RED metrics + `search_view_rows` + `search_projection_lag_messages`.", "responses": { "200": { "description": "text/plain; version=0.0.4" } } } },
            "/api/v1/search/products": { "get": { "operationId": "searchProducts", "tags": ["products"], "summary": "Product search", "description": "Bilingual full-text (tsvector) + trigram fallback, category filter, price-ordered, with category facets and pagination.", "parameters": [
                { "name": "q", "in": "query", "description": "search terms (English/Bangla)", "schema": { "type": "string" }, "example": "rice" },
                { "name": "locale", "in": "query", "description": "result locale; one of en|bn", "schema": { "type": "string", "enum": ["en", "bn"], "default": "en" } },
                { "name": "category_id", "in": "query", "description": "filter by category UUID", "schema": { "type": "string", "format": "uuid" } },
                { "name": "page", "in": "query", "description": "1-based page", "schema": { "type": "integer", "minimum": 1, "default": 1 } },
                { "name": "size", "in": "query", "description": "page size (1..100)", "schema": { "type": "integer", "minimum": 1, "maximum": 100, "default": 20 } }
            ], "responses": { "200": json_resp("results + facets", "ProductSearchResponse"), "422": json_resp("invalid_request (locale/size/category_id)", "ErrorEnvelope") } } },
            "/api/v1/search/autocomplete": { "get": { "operationId": "autocompleteProducts", "tags": ["products"], "summary": "Autocomplete suggestions", "description": "Trigram similarity over product names; returns up to 10 suggestions.", "parameters": [
                { "name": "q", "in": "query", "required": true, "description": "prefix/partial term", "schema": { "type": "string" }, "example": "ri" },
                { "name": "locale", "in": "query", "schema": { "type": "string", "enum": ["en", "bn"], "default": "en" } }
            ], "responses": { "200": json_resp("suggestions", "AutocompleteResponse"), "422": json_resp("q required / bad locale", "ErrorEnvelope") } } },
            "/api/v1/search/shops": { "get": { "operationId": "searchShops", "tags": ["shops"], "summary": "Nearby shops", "description": "Geo radius search via PostgreSQL earthdistance; ordered by distance.", "parameters": [
                { "name": "lat", "in": "query", "required": true, "description": "latitude (-90..90)", "schema": { "type": "number", "format": "double" }, "example": 23.8103 },
                { "name": "lng", "in": "query", "required": true, "description": "longitude (-180..180)", "schema": { "type": "number", "format": "double" }, "example": 90.4125 },
                { "name": "radius_km", "in": "query", "description": "search radius (1..30)", "schema": { "type": "integer", "minimum": 1, "maximum": 30, "default": 5 } },
                { "name": "q", "in": "query", "description": "optional name filter", "schema": { "type": "string" } }
            ], "responses": { "200": json_resp("shops by distance", "ShopSearchResponse"), "422": json_resp("lat/lng required or out of range", "ErrorEnvelope") } } },
            "/api/v1/search/trending": { "get": { "operationId": "getTrending", "tags": ["discovery"], "summary": "Trending products", "description": "Top products by order volume over the trailing 7 days.", "responses": { "200": json_resp("trending items", "TrendingResponse") } } },
            "/api/v1/search/categories/tree": { "get": { "operationId": "getCategoryTree", "tags": ["discovery"], "summary": "Category tree", "description": "Active categories as a 2-level tree (roots + children).", "responses": { "200": json_resp("category tree", "CategoryTreeResponse") } } },
            "/api/v1/search/admin/reindex": { "post": { "operationId": "reindexSearch", "tags": ["admin"], "summary": "Admin reindex", "description": "Schedules a re-projection. Requires a valid RS256 token with the admin role.", "security": [{ "bearerJwt": [] }], "responses": {
                "202": json_resp("accepted", "ReindexResponse"), "401": json_resp("missing/invalid token", "ErrorEnvelope"), "403": json_resp("not an admin", "ErrorEnvelope") } } }
        },
        "components": {
            "securitySchemes": { "bearerJwt": { "type": "http", "scheme": "bearer", "bearerFormat": "JWT", "description": "RS256 access token minted by 01-auth." } },
            "schemas": {
                "Identity": { "type": "object", "properties": {
                    "service_name": { "type": "string", "example": "05-search" }, "code_version": { "type": "string" },
                    "env_version": { "type": "string" }, "tenant": { "type": "string", "enum": ["local", "cloud"] },
                    "env": { "type": "string" }, "uptime_seconds": { "type": "integer" } } },
                "ErrorEnvelope": { "type": "object", "properties": { "error": { "type": "object", "properties": {
                    "code": { "type": "string", "example": "invalid_request" }, "message": { "type": "string" },
                    "request_id": { "type": "string" }, "details": { "type": "object", "additionalProperties": true } } } } },
                "Dependency": { "type": "object", "properties": { "name": { "type": "string" }, "reachable": { "type": "boolean" }, "latency_ms": { "type": "number" }, "detail": { "type": "string" } } },
                "ReadyResponse": { "type": "object", "properties": { "status": { "type": "string", "enum": ["ready", "not_ready"] }, "identity": ref_("Identity"), "dependencies": { "type": "array", "items": ref_("Dependency") } } },
                "Check": { "type": "object", "properties": { "ok": { "type": "boolean" }, "latency_ms": { "type": "number" }, "detail": { "type": "string" } } },
                "HealthResponse": { "type": "object", "properties": {
                    "status": { "type": "string", "enum": ["healthy", "unhealthy"] }, "identity": ref_("Identity"),
                    "checks": { "type": "object", "properties": { "postgres": ref_("Check"), "elasticsearch": ref_("Check"), "log_elasticsearch": ref_("Check"), "kafka": ref_("Check"), "mongo_logs": ref_("Check"), "apm": ref_("Check") } },
                    "projection": { "type": "object", "properties": { "products_view": { "type": "integer" }, "shops_view": { "type": "integer" }, "categories_view": { "type": "integer" } } },
                    "observability": { "type": "object", "properties": { "apm_service_name": { "type": "string" }, "logs_sink_es": { "type": "string" }, "logs_sink_mongo": { "type": "string" }, "search_es": { "type": "string" } } } } },
                "ProductItem": { "type": "object", "properties": {
                    "product_id": { "type": "string", "format": "uuid" }, "shop_id": { "type": "string", "format": "uuid", "nullable": true },
                    "category_id": { "type": "string", "format": "uuid", "nullable": true }, "name_en": { "type": "string" }, "name_bn": { "type": "string" },
                    "slug": { "type": "string", "nullable": true }, "list_price_minor": { "type": "integer", "description": "price in paisa" },
                    "sale_price_minor": { "type": "integer", "nullable": true }, "rating_avg": { "type": "number" }, "rating_count": { "type": "integer" }, "in_stock": { "type": "boolean" } } },
                "Facets": { "type": "object", "properties": { "category": { "type": "object", "additionalProperties": { "type": "integer" }, "description": "category_id → count" } } },
                "ProductSearchResponse": { "type": "object", "properties": {
                    "total": { "type": "integer" }, "page": { "type": "integer" }, "size": { "type": "integer" }, "locale": { "type": "string" },
                    "items": { "type": "array", "items": ref_("ProductItem") }, "facets": ref_("Facets") } },
                "Suggestion": { "type": "object", "properties": { "text": { "type": "string" }, "product_id": { "type": "string", "format": "uuid" } } },
                "AutocompleteResponse": { "type": "object", "properties": { "suggestions": { "type": "array", "items": ref_("Suggestion") } } },
                "ShopItem": { "type": "object", "properties": {
                    "shop_id": { "type": "string", "format": "uuid" }, "handle": { "type": "string" }, "name_en": { "type": "string" }, "name_bn": { "type": "string" },
                    "lat": { "type": "number" }, "lng": { "type": "number" }, "distance_km": { "type": "number" }, "rating_avg": { "type": "number" }, "rating_count": { "type": "integer" }, "open_now": { "type": "boolean" } } },
                "ShopSearchResponse": { "type": "object", "properties": { "total": { "type": "integer" }, "items": { "type": "array", "items": ref_("ShopItem") } } },
                "TrendingItem": { "type": "object", "properties": { "product_id": { "type": "string", "format": "uuid" }, "name_en": { "type": "string" }, "name_bn": { "type": "string" }, "orders": { "type": "integer" }, "list_price_minor": { "type": "integer" }, "rating_avg": { "type": "number" } } },
                "TrendingResponse": { "type": "object", "properties": { "items": { "type": "array", "items": ref_("TrendingItem") } } },
                "CategoryChild": { "type": "object", "properties": { "category_id": { "type": "string", "format": "uuid" }, "name_en": { "type": "string" }, "name_bn": { "type": "string" }, "slug": { "type": "string", "nullable": true } } },
                "CategoryNode": { "type": "object", "properties": { "category_id": { "type": "string", "format": "uuid" }, "name_en": { "type": "string" }, "name_bn": { "type": "string" }, "slug": { "type": "string", "nullable": true }, "parent_id": { "type": "string", "format": "uuid", "nullable": true }, "children": { "type": "array", "items": ref_("CategoryChild") } } },
                "CategoryTreeResponse": { "type": "object", "properties": { "tree": { "type": "array", "items": ref_("CategoryNode") } } },
                "ReindexResponse": { "type": "object", "properties": { "job_id": { "type": "string", "format": "uuid" }, "status": { "type": "string", "example": "accepted" }, "message": { "type": "string" } } }
            }
        }
    })
}

pub const SWAGGER_HTML: &str = r##"<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"><title>05-search API</title><link rel="stylesheet" href="https://unpkg.com/swagger-ui-dist@5/swagger-ui.css"></head><body><div id="swagger-ui"></div><script src="https://unpkg.com/swagger-ui-dist@5/swagger-ui-bundle.js"></script><script>window.ui=SwaggerUIBundle({url:"/openapi.json",dom_id:"#swagger-ui",deepLinking:true,tryItOutEnabled:true,docExpansion:"list",defaultModelsExpandDepth:2,persistAuthorization:true});</script></body></html>"##;
